### PR TITLE
Symmetric tensor cleanup

### PR DESCRIPTION
**Pull request by drwells**
I'm using `SymmetricTensor` for some solid mechanics operations and I noticed (via callgrind) that this loop

https://github.com/dealii/dealii/blob/8d0257005b44273c51e31a0c8fa08e511ec7e245/include/deal.II/base/symmetric_tensor.h#L2038-L2041

is inlined but run multiple times per SymmetricTensor-Tensor multiplication. In the process of optimizing this I found a bunch of things that can be cleaned up in SymmetricTensor; e.g., we have three different copies of that loop in `symmetric_tensor.h`: this PR gets rid of two of them (and uses the version with index lookup tables for small tensors).

### CODE REVIEW

File: include/deal.II/base/symmetric_tensor.h [hunk 1]
```diff
@@ -181,49 +181,6 @@ determinant(const SymmetricTensor<2, dim, Number> &);
 
 namespace internal
 {
-  // Workaround: The following 4 overloads are necessary to be able to
```
Comment (blaisb) [inline include/deal.II/base/symmetric_tensor.h:184]: These are not necessary anymore I presume?

File: include/deal.II/base/symmetric_tensor.h [hunk 2]
```diff
@@ -2005,49 +1962,96 @@ DEAL_II_HOST DEAL_II_CONSTEXPR inline
...
+                return indices[0];
+
+              const TableIndices<2> sorted_indices(
+                std::min(indices[0], indices[1]),
+                std::max(indices[0], indices[1]));
+
+              for (unsigned int d = 0, c = 0; d < dim; ++d)
+                for (unsigned int e = d + 1; e < dim; ++e, ++c)
+                  if ((sorted_indices[0] == d) && (sorted_indices[1] == e))
+                    return dim + c;
```
Comment (blaisb) [inline include/deal.II/base/symmetric_tensor.h:2022]: I get what you are doing here, but could you add a bit of comments that explain it in more detail? Especially the trick with c=0 being declared in line 2019 but incremented in 2020

**Review by blaisb — CHANGES_REQUESTED** [review]
Very few minor comments. This is a really clean implementation. Does it impact positively the performance?

File: include/deal.II/base/symmetric_tensor.h (hunk 1 above)
Comment (drwells) [reply include/deal.II/base/symmetric_tensor.h]: While we don't support those compilers any more these overloads are necessary for nvcc - I'm going to leave them back in for now.

File: include/deal.II/base/symmetric_tensor.h (hunk 2 above)
Comment (drwells) [reply include/deal.II/base/symmetric_tensor.h]: Fixed

**Comment by drwells** [comment]
Thanks.

> Does it impact positively the performance?

I'll check

**Review by kronbichler — COMMENTED** [review]
This looks good, it is great to unify the code. I can imagine that this could have lead to bad compiled code, because it is hard to figure out for the compiler if it only partially unrolls the loops or does not run the optimization passes in the best order.

**Review by kronbichler — APPROVED** [review]

**Comment by drwells** [comment]
I did some benchmarking [1] and things are marginally faster (about 1311ms vs 1324ms) with this patch. Things are faster yet if I manually unroll the loop in `operator*(const SymmetricTensor<2, 3> &, const Tensor<1, 3> &)` - I want to profile that some more so that will go in another PR.

[1] source:
```cpp
#include <deal.II/base/symmetric_tensor.h>

#include <chrono>
#include <iostream>

using namespace dealii;

template <typename T>
Tensor<1, 3>
faster_vmult(const T &A, const Tensor<1, 3, double> &x)
{
  Tensor<1, 3> result;
  result[0] = A[0][0] * x[0] + A[0][1] * x[1] + A[0][2] * x[2];
  result[1] = A[1][0] * x[0] + A[1][1] * x[1] + A[1][2] * x[2];
  result[2] = A[2][0] * x[0] + A[2][1] * x[1] + A[2][2] * x[2];

  return result;
}

Tensor<1, 3>
faster_vmult(const SymmetricTensor<2, 3, double> &A, const Tensor<1, 3, double> &x)
{
  Tensor<1, 3> result;
  const double A00 = A.access_raw_entry(0);
  const double A01 = A.access_raw_entry(3);
  const double A02 = A.access_raw_entry(4);
  const double A10 = A01;
  const double A11 = A.access_raw_entry(1);
  const double A12 = A.access_raw_entry(5);
  const double A20 = A02;
  const double A21 = A12;
  const double A22 = A.access_raw_entry(2);
  result[0] = A00 * x[0] + A01 * x[1] + A02 * x[2];
  result[1] = A10 * x[0] + A11 * x[1] + A12 * x[2];
  result[2] = A20 * x[0] + A21 * x[1] + A22 * x[2];

  return result;
}

template <typename T>
double
vmult_and_sum(const T &A, const Tensor<1, 3, double> &x)
{
  const auto result  = A * x;
  const auto result1 = A * result;
  const auto result2 = A * result1;
  // const auto result = faster_vmult(A, x);
  // const auto result1 = faster_vmult(A, result);
  // const auto result2 = faster_vmult(A, result1);
  return result[0] + result1[1] + result2[2];
}

int
main()
{
  // pre-cleanup:
  // Conclusion 1: for operator*: SymmetricTensor is faster on archway (1324ms vs 1404 ms)
  // Conclusion 2: for faster_vmult: Tensor is faster on archway (1081ms vs 1153ms)
  // post-cleanup:
  // Conclusion 3: for operator*: SymmetricTensor: 1311ms
  // Conclusion 4: for faster_vmult: SymmetricTensor: 1081ms
  SymmetricTensor<2, 3, double> A;
  for (unsigned int i = 0; i < 3; ++i)
    for (unsigned int j = 0; j < 3; ++j)
      A[i][j] = double(i + j);

  Tensor<1, 3, double> x;
  x[0]          = 1.0;
  x[1]          = 2.0;
  x[2]          = 0.0;
  const auto t0 = std::chrono::high_resolution_clock::now();
  for (std::size_t i = 0; i < 100000000; ++i)
    {
      x[0] += vmult_and_sum(A, x);
      x[0] *= 0.0001;
    }
  const auto t1 = std::chrono::high_resolution_clock::now();

  std::cout << "time elapsed: "
            << std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count()
            << '\n';

  return std::size_t(x[0]) % 8;
}
```

**Comment by kronbichler** [comment]
I am intrigued by your numbers @drwells. I think the main problem must be that https://github.com/dealii/dealii/blob/8d0257005b44273c51e31a0c8fa08e511ec7e245/include/deal.II/base/symmetric_tensor.h#L4208-L4212 is ill-written, because we force the compiler to add into a zeroed-out field. One could wish that the compiler unrolled the loops, but unfortunately that does not work due to floating-point rules (I believe it is subnormals or behavior near infinity, but I don't recall the exact rule). Let me open a PR for that in a minute. The other thing is that the benchmark is a bit too simple (in my humble opinion) because the compiler can use that some entries in the vector `x` are loop invariants and some entries are zero. Without knowing what precisely should get benchmarked, I would have at least used
```
  double input = 0.0;
  if (argc > 1)
    input = std::atof(argv[1]);

  SymmetricTensor<2, 3, double> A;
  for (unsigned int i = 0; i < 3; ++i)
    for (unsigned int j = 0; j < 3; ++j)
      A[i][j] = double(i + j) + input;

  Tensor<1, 3, double> x;
  x[0]          = 1.0;
  x[1]          = 2.0;
  x[2]          = input;
  const auto t0 = std::chrono::high_resolution_clock::now();
  for (std::size_t i = 0; i < 100000000; ++i)
    {
      x[0] += vmult_and_sum(A, x);
      x[0] *= 0.0001;
    }
  const auto t1 = std::chrono::high_resolution_clock::now();
```
as the benchmark body, because then the compiler can't optimize for zero.

**Comment by drwells** [comment]
Those are good points - I like the `atof(argv[1])` trick.

I have another question or two about this but I'll ask them over in the other PR.

**Review by blaisb — APPROVED** [review]

**Pull request merged**